daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/ExpressionEvaluatingUnparsers.scala

```diff
@@ -77,30 +78,83 @@ final class SetVariableUnparser(
 
 }
 
+final class NewVariableInstanceSuspendableExpression(
+  override val expr: CompiledExpression[AnyRef],
+  override val rd: VariableRuntimeData)
+  extends SuspendableExpression {
+
+  override protected def processExpressionResult(ustate: UState, v: DataValuePrimitive): Unit = {
+    val vi = ustate.variableMap.find(rd.globalQName)
+    Assert.invariant(vi.isDefined)
+    // Reset original variable's state from VariableInProcess
+    vi.get.setState(vi.get.priorState)
+    ustate.variableMap.newVariableInstance(rd, v)
+  }
+
+  override protected def maybeKnownLengthInBits(ustate: UState) = MaybeULong(0)
+}
+
 // When implemented this almost certainly wants to be a combinator
 // Not two separate unparsers.
-class NewVariableInstanceStartUnparser(override val context: RuntimeData)
+class NewVariableInstanceStartUnparser(override val context: VariableRuntimeData)
   extends PrimUnparserNoData {
 
   override lazy val runtimeDependencies = Vector()
 
   override lazy val childProcessors = Vector()
 
   override def unparse(state: UState) = {
-    val vrd = context.asInstanceOf[VariableRuntimeData]
-    state.newVariableInstance(vrd)
+    if (context.maybeDefaultValueExpr.isDefined) {
+      val maybeVar = state.variableMap.find(context.globalQName)
+      Assert.invariant(maybeVar.isDefined)
+
+      // Set the variable's state to VariableInProcess while waiting for the NVI
+      // expression to complete. When the expression completes, this original
+      // variable instance will have its state reset to its original value.
+      maybeVar.get.setState(VariableInProcess)
+      val suspendableExpression = new NewVariableInstanceSuspendableExpression(context.maybeDefaultValueExpr.get, context)
+      suspendableExpression.run(state)
+    } else
+      state.variableMap.newVariableInstance(context)
   }
 }
 
-class NewVariableInstanceEndUnparser(override val context: RuntimeData)
-  extends PrimUnparserNoData {
+final class NewVariableInstanceSuspendableOperation(override val rd: VariableRuntimeData)
+  extends SuspendableOperation {
+
+    override val isReadOnly = true
+
+    override def toString = "Suspension for removal of variable instance: " + rd.globalQName
```

Comment:
I'd prefer we just stick with the default toString for suspendable operations so that the strings are consistent. I'm working on a debugger patch to output suspension information and this string is going to stand out as very different.